internal/testing/cmdtest/cmdtest.go

```diff
@@ -0,0 +1,610 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// The cmdtest package simplifies testing of command-line interfaces. It
+// provides a simple, cross-platform, shell-like language to express command
+// execution. It can compare actual output with the expected output in the file,
+// and can also update a file with new "golden" output that is deemed correct.
+//
+// Start using cmdtest by writing a test file with commands and expected output.
+// See the TestFile documentation for the syntax.
+//
+// To test, first read the file:
+//
+//    tf, err := cmdtest.Read("test.ct")
+//
+// Then configure the resulting TestFile by adding commands or enabling
```

Comment:
I didn't see any example of this (doing anything with `TestFile` other than `Compare` or `Update`). Is there a use case?

Related: would it be simpler to have `cmdtest` define the `--record` flag, and just have the interface to be something more like "run the tests in this file", and let it manage reading/writing the expected output based on `--record`?

Reply:
See the branch in my fork for examples.

It's bad practice in Go for packages to define flags. A package should have a purely programmatic API. (Also, `-record` is the wrong name; it should be `-update` or something like that, IMO.)

Reply:
Maybe you have a better sense of this than I do at this point, but it still feels like you're requiring the caller to do more than necessary (i.e., there will be a lot of repeated code in callers).

Can't use of this package look something like this?

```
record := false // usually would be defined as a flag
tf := cmdtest.New(testDataDir)
tf.Setup := func... // optional
tf.AddBinary(myBinaryName, pathToMyBinaryExecutable) // optional
tf.AddCommand("foo", func() { whatever }) // optional
if err := tf.Run(record); err != nil {
  t.Fatal(err)
}
```

Reply:
Added `Run`. Left `Compare` and `Update`; should I unexport them?

Reply:
IIUC, you did about half of what I was suggesting -- you combined `Compare`/`Update` into `Run`, but you didn't respond to the part about changing `cmdtest` to operate on a directory. Note that in my pseudocode, you provide a directory, configuration (once), and then `Run` does the work of finding tests, `Read`ing them, etc.

This seems noticeably easier to me than having every client do the work of iterating over all the tests. It also feels weird to me to have to re-configure each `TestFile` (e.g., to add `Command`s).

Is there a reason to keep the interface as a single test rather than as a test suite like that?

Reply:
Sorry, I missed all that. Makes sense.

Done. I kept `Commands` as an exported map, and I use `Read` instead of `New` (because I think of `New` as an in-memory thing), but otherwise it matches your snippet.

PTAL